### ISSUE TITLE
USB:storage: Convoluted PUN_INFO handling

### ISSUE THREAD

**Issue by czietz** [freemint/freemint#323]
Perhaps I'm missing something, but the handling of the PUN_INFO structure (that carries information linking logical drives to physical units) seems very convoluted in the USB storage driver.

There is the private structure `pun_usb`, which is setup like a PUN_INFO (with AHDI cookie, fake AHDI version number etc.) but apparently never exposed via the `pun_ptr` system variable (memory address 0x516):
https://github.com/freemint/freemint/blob/a9fd0ebd8d0991b5d76c17b99ac667aad0592266/sys/usb/src.km/udd/storage/usb_storage.c#L2530-L2539

Then there is `pun_ahdi`, which is never filled with information about the USB devices as far as I can see, but is installed into `pun_ptr` under certain circumstances:
https://github.com/freemint/freemint/blob/a9fd0ebd8d0991b5d76c17b99ac667aad0592266/sys/usb/src.km/udd/storage/usb_storage.c#L2377

And, more importantly, in case `pun_ahdi` is installed into `pun_ptr`, the contents of the previous PUN_INFO structure are not copied over, so that the information setup by the hard disk driver is not available anymore to applications following `pun_ptr`.
https://github.com/freemint/freemint/blob/a9fd0ebd8d0991b5d76c17b99ac667aad0592266/sys/usb/src.km/udd/storage/usb_storage.c#L2389-L2447

**Comment by DavidGZ** [comment]
Hi Christian, I can take a look but not before Christmas vacation.

**Comment by DavidGZ** [comment]
> Perhaps I'm missing something, but the handling of the PUN_INFO structure (that carries information linking logical drives to physical units) seems very convoluted in the USB storage driver. 
> 

Yes, it's strange how PUN_INFO is handled, if I remember correctly, although different people have worked on this driver, how this is handled comes from Didier's TOS sources for the FireBee and the CTPCI. I remember guessing that Didier had his reasons to do it like this and because it was working and there were many things to do regarding USB I didn't investigate this further when I imported his sources to FreeMiNT.

> There is the private structure `pun_usb`, which is setup like a PUN_INFO (with AHDI cookie, fake AHDI version number etc.) but apparently never exposed via the `pun_ptr` system variable (memory address 0x516):
> 

This structure used internally is an extended version of the documented PUN_INFO, I think the new members of the struct were Didier's additions. Also it handles 32 logical devices instead of 16.

https://github.com/freemint/freemint/blob/ccaa16937de9aa4bc67d1c5de40f93236dd3053e/sys/usb/src.km/udd/storage/xhdi.h#L61-L64

> 
> Then there is `pun_ahdi`, which is never filled with information about the USB devices as far as I can see, but is installed into `pun_ptr` under certain circumstances:
> 

> 
> And, more importantly, in case `pun_ahdi` is installed into `pun_ptr`, the contents of the previous PUN_INFO structure are not copied over, so that the information setup by the hard disk driver is not available anymore to applications following `pun_ptr`.
> 

As far as I can see pun_ahdi is only installed when pun_ptr doesn't point to any PUN_INFO structure, so no hard disk driver was previously present.

I don't know if PUN_INFO was thought to be handled by two different drivers, according to XHDI documentation it wasn't, and I'm not sure if we can just replace the PUN_INFO pointed by pun_ptr by a new undocumented version of it without causing troubles with the already installed hard disk driver. I think this is  why Didier decided that the USB mass storage driver shouldn't touch the PUN_INFO structure pointed by pun_ptr.

**Comment by czietz** [comment]
Maybe I wasn't clear enough in my original issue description. What happens is as follows:

- HDDRIVER installs a PUN_INFO structure and puts the pointer into 0x516.
- The USB storage driver decides to install a _new_ PUN_INFO structure into 0x516, without even copying the previous entries.
- Access to the information placed by HDDRIVER into its PUN_INFO structure is lost.

This is a bug. At the very least install_pun() needs to preserve information from the previous PUN_INFO.

**Comment by DavidGZ** [comment]
OK, I see, I've misread the code in install_pun().

**Comment by DavidGZ** [comment]
For now I have stopped the driver installing a new PUN_INFO when there is another one already installed.
https://github.com/freemint/freemint/commit/bc6492f7a9a50357cc75ddb12494882bcc4afe4a

Perhaps we could try to improve further the PUIN_INFO handling to make it less confusing.

I've gone over the history of this code and I was wrong in my comment above regarding Didier's code not touching the PUN_INFO struct for USB devices, in fact his code was updating PUN_INFO with the new USB devices, but it was limited only to 16 logical devices.

**Comment by czietz** [comment]
> For now I have stopped the driver installing a new PUN_INFO when there is another one already installed. 

Thank you. I can only test next year, but looking at the code I assume it should fix the issue I observed.

> I've gone over the history of this code and I was wrong in my comment above regarding Didier's code not touching the PUN_INFO struct for USB devices, in fact his code was updating PUN_INFO with the new USB devices, but it was limited only to 16 logical devices.

It might be a worthwhile addition to update the public PUN_INFO structure with information about the USB devices. After all, there might be software that relies on the information in that structure.

**Comment by DavidGZ** [comment]
> It might be a worthwhile addition to update the public PUN_INFO structure with information about the USB devices. After all, there might be software that relies on the information in that structure.

Now the driver updates public PUN_INFO structure with USB information.
https://github.com/freemint/freemint/commit/287b69264ee230a5a6b62d399ecc17e1cf91377b

**Comment by fdanapfel** [comment]
I already reported the following at https://www.atari-forum.com/viewtopic.php?p=458645#p458645 but was advised to add it to this issue, since the issue I'm having seems to be related to the code changes discussed here.

These changes seem to cause issues when using an USB stick on a NetUSBee connected to an Atari TT030 running TOS 3.06 and using HDDRIVER 12.03.

Plugging in the USB stick and reading and writing to it works as expected, but when the USB stick is unplugged the Atari hangs with 2 Bombs (the mouse pointer still works).

When switching back to the snapshot of the drivers from December 19th, 2023 (usb4tos-1-19-ccaa1693.zip) unplugging the USB stick works again.

**Comment by fdanapfel** [comment]
@czietz 
> Could you please check whether commit [bc6492f](https://github.com/freemint/freemint/commit/bc6492f7a9a50357cc75ddb12494882bcc4afe4a) or commit [287b692](https://github.com/freemint/freemint/commit/287b69264ee230a5a6b62d399ecc17e1cf91377b) is first non-working version? I see a potential problem affecting USB stick removal, which was introduced in commit [287b692](https://github.com/freemint/freemint/commit/287b69264ee230a5a6b62d399ecc17e1cf91377b), dated December 27th:

Yes, commit [287b692](https://github.com/freemint/freemint/commit/287b69264ee230a5a6b62d399ecc17e1cf91377b) appears to be causing the issue.

With snapshot [usb4tos-1-19-bc6492f7.zip](https://tho-otto.de/snapshots/freemint/usb4tos/usb4tos-1-19-bc6492f7.zip) from December 25th unplugging the USB stick works, but when using snapshot [usb4tos-1-19-287b6926.zip](https://tho-otto.de/snapshots/freemint/usb4tos/usb4tos-1-19-287b6926.zip) from December 27th I get the 2 bombs when the USB stick is unplugged.

**Comment by czietz** [comment]
> Yes, commit [287b692](https://github.com/freemint/freemint/commit/287b69264ee230a5a6b62d399ecc17e1cf91377b) appears to be causing the issue.

Thank you for your report and for confirming the breaking commit. I have just pushed a fix with commit 6f6376da. Unfortunately, I cannot test myself right now. But if you want to give it a try: the new drivers are already available at https://tho-otto.de/snapshots/freemint/usb4tos/.

**Comment by fdanapfel** [comment]
> Thank you for your report and for confirming the breaking commit. I have just pushed a fix with commit [6f6376d](https://github.com/freemint/freemint/commit/6f6376da34f938c0c42fed602091c3bade0392ce). Unfortunately, I cannot test myself right now. But if you want to give it a try: the new drivers are already available at https://tho-otto.de/snapshots/freemint/usb4tos/.

Thanks for the fix. I can confilrm that wit snapshot [usb4tos-1-19-6f6376da.zip](https://tho-otto.de/snapshots/freemint/usb4tos/usb4tos-1-19-6f6376da.zip) from today unplugging the USB Stick works again without Bombs.

**Comment by DavidGZ** [comment]
@czietz thanks for taking care of the issue

**Comment by mikrosk** [comment]
It seems that all issues have been addressed, closing.
